### ISSUE TITLE
Idea: EOL "apply" in favor of more descriptive verbs

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#118775]
https://github.com/kubernetes/kubernetes/issues/118725 illustrates some surprising behavior with `kubectl apply`.  I have found, repeatedly, that humans are bad at predicting what apply will do (made worse by client-side and server-side being different).

From talking to a lot of people, the `apply` action feels like it means `create-or-replace`, but it's really not.  It's a much more sophisticated resource co-ownership system.  If you don't keep that in mind, it will almost certainly NOT do what you expect it to do, which has led to many bugs over the years (as @apelisse can attest).

In hindsight, I wonder if "apply" was the right name for this.  Perhaps we really want 2 distinct operations - something like "kubectl set" (simple create-or-replace) and "kubectl merge" (what we call "apply" today).  I suspect that vast majority of issues reported with "apply" would be avoided because humans would mostly use "set", and "merge" would be much more clearly a variant of patch, left for "advanced" cases.

I wonder if it is possible to steer towards a future where this is clearer.  @apelisse @soltysh @KnVerey @eddiezane @jpbetz

**Comment by apelisse** [comment]
I totally agree with a lot of the sentiment here. I have regularly people asking me why `kubectl apply` doesn't reset some of the fields that they haven't specified when they apply. I don't have field XYZ in my manifest, and when I apply it, it doesn't remove field XYZ, why? Well, because we're merging, not "replacing". But then, I ask, do you want to remove the status? probably not, do you want to remove clusterIP? well that doesn't work. So I don't know if people want "merge" or "replace", I think people just don't know what they want? I'm still not sure I know what they want either.

**Comment by thockin** [comment]
I think spec vs status is OK.  But it gets way more subtle, like the service-ports merging.  It's a correct result for "merge" but very surprising for "apply" and would be obviously wrong for "set".

**Comment by apelisse** [comment]
Hard for me to understand what you mean by that since I can use "apply" and "merge" almost indistinctly.
Apply has always meant "merge" (see #1702), and as an implementation, has always used strategic-merge-patch, and/or json-merge-patch, both merging mechanisms. The term "apply", in my understanding, means "apply these changes on top of the existing resource", with the extra-feature of removing things that I used to have but don't have anymore. People can already replace objects using PUT, but I have a feeling that won't make solve their problem either.

**Comment by thockin** [comment]
I mean: for many cases, "apply" and "replace" are functionally the same (or close enough).  Thus, many people think of "apply" as "fancy replace", when it's really "fancy patch".  That leads to surprises and frustration, and bug reports that are actually WAI.

**Comment by thockin** [comment]
Also "terraform apply" is closer to "fancy replace" than "fancy patch", so using the same word causes friction.
